### PR TITLE
Replace PTZ boolean normalization with boolean constraints

### LINKED ISSUE #256: Clarify meaning of PTZ constraints presence
This spec [appears to assume](https://w3c.github.io/mediacapture-image/#pan) any mention of `pan`, `tilt` or `zoom` will result in a PTZ device (& thus require ptz permission).

But that's not how the constraints algorithm works. E.g.
```js
navigator.mediaDevices.getUserMedia({video: {height: {min: 1080}, pan: true, tilt: true, zoom: true}});
```
...is equivalent to 
```js
navigator.mediaDevices.getUserMedia({video: {height: {min: 1080}, pan: {}, tilt: {}, zoom: {}}});
```
...where `{}` has no impact on [fitness distance](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-fitness-distance), and may choose your 1080p camera over your low-res PTZ camera.  

This seems unexpected, because a constraint value (even `true`) would usually carry weight, except as defined here.

You may also have granted the site more permission than necessary to access the device chosen.

### What's the model?

A related question is whether regular non-ptz cameras satisfy the following constraints:
```js
navigator.mediaDevices.getUserMedia({video: {pan: 0, tilt: 0, zoom: 0}});
```
Is it yes, because non-ptz cameras point front and center without zoom or tilt?
Is it no, because non-ptz cameras shouldn't have read-only `pan`, `tilt` or `zoom` settings? The spec doesn't say.

It appears we've overloaded the presence or absence of `pan`, `tilt` or `zoom` constraints with meaning.

### Options I see:
 1. Remove the `true` overload of `pan`, `tilt` and `zoom` constraints, and replace it with a `ptz` [ConstrainBoolean](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-constrainboolean).
 2. Hotfix the fitness distance algorithm to give weight to presence of constraints.
 3. Make `true` a first-class value of `pan`, `tilt` and `zoom`, and specify a new `ConstrainULongOrBoolean`.

### PR DESCRIPTION
**Pull request by eehakkin**
Fixes #256. This along with https://github.com/w3c/mediacapture-main/pull/742 is to clarify meaning of PTZ constraints presence.

***
<a href="https://pr-preview.s3.amazonaws.com/eehakkin/intel-w3c-mediacapture-image/pull/271.html" title="Last updated on Mar 16, 2021, 2:40 PM UTC (3bef8ed)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/mediacapture-image/271/72ee760...eehakkin:3bef8ed.html" title="Last updated on Mar 16, 2021, 2:40 PM UTC (3bef8ed)">Diff</a>

### CODE REVIEW

**Comment by beaufortfrancois** [comment]
@jan-ivar @guidou @youennf Does this look good to you?

File: index.bs [hunk 1]
```diff
@@ -795,20 +789,19 @@ When the {{getSettings()}} method is invoked on a video stream track, the user a
...
 
-    Any algorithm which uses a {{MediaTrackConstraintSet}} object and its {{MediaTrackConstraintSet/pan}} dictionary member which exists after a possible normalization MUST <a>request permission to use</a> (as defined in [[!permissions]]) a PermissionDescriptor with its name member set to {{PermissionName/camera}} and its {{CameraDevicePermissionDescriptor/panTiltZoom}} member set to true, and, optionally, consider its {{DevicePermissionDescriptor/deviceId}} member set to any appropriate device's deviceId.
+    Pan is also a boolean constrainable property corresponding to the camera's ability to pan (which results in a proper non-degenerate numeric setting range) and to the user agent's decision to expose that ability.
+    This boolean constraint is satisfied by a <a>settings dictionary</a> if the constraint value matches the existence of the numeric pan setting in the settings dictionary.
```
Comment (jan-ivar) [inline index.bs:793]: The _constrainable property_ is never a boolean, only the constraint. Once https://github.com/w3c/mediacapture-main/pull/766 is merged, we can say:
```suggestion
    Constraints on pan influence camera selection through {{fitness distance}} toward cameras with the ability to pan. To exert this influence without overwriting the current pan setting, pan may be constrained to true. Conversely, constraining it to false disfavors cameras with the ability to pan.
```

File: index.bs [hunk 2]
```diff
@@ -795,20 +789,19 @@ When the {{getSettings()}} method is invoked on a video stream track, the user a
...
 
-    Any algorithm which uses a {{MediaTrackConstraintSet}} object and its {{MediaTrackConstraintSet/tilt}} dictionary member which exists after a possible normalization MUST <a>request permission to use</a> (as defined in [[!permissions]]) a PermissionDescriptor with its name member set to {{PermissionName/camera}} and its {{CameraDevicePermissionDescriptor/panTiltZoom}} member set to true, and, optionally, consider its {{DevicePermissionDescriptor/deviceId}} member set to any appropriate device's deviceId.
+    Tilt is also a boolean constrainable property corresponding to the camera's ability to tilt (which results in a proper non-degenerate numeric setting range) and to the user agent's decision to expose that ability.
+    This boolean constraint is satisfied by a <a>settings dictionary</a> if the constraint value matches the existence of the numeric tilt setting in the settings dictionary.
```
Comment (jan-ivar) [inline index.bs:802]: adding "with a value other than false" and a comma
```suggestion
    Constraints on tilt influence camera selection through {{fitness distance}} toward cameras with the ability to tilt. To exert this influence without overwriting the current tilt setting, tilt may be constrained to true. Conversely, constraining it to false disfavors cameras with the ability to tilt.
```

File: index.bs [hunk 3]
```diff
@@ -818,10 +811,12 @@ When the {{getSettings()}} method is invoked on a video stream track, the user a
 
     <li><dfn>Zoom</dfn> is a numeric camera setting that controls the focal length of the lens. The setting usually represents a ratio, e.g. 4 is a zoom ratio of 4:1. The minimum value is usually 1, to represent a 1:1 ratio (i.e. no zoom).
 
-    Any algorithm which uses a {{MediaTrackConstraintSet}} object and its {{MediaTrackConstraintSet/zoom}} dictionary member which exists after a possible normalization MUST <a>request permission to use</a> (as defined in [[!permissions]]) a PermissionDescriptor with its name member set to {{PermissionName/camera}} and its {{CameraDevicePermissionDescriptor/panTiltZoom}} member set to true, and, optionally, consider its {{DevicePermissionDescriptor/deviceId}} member set to any appropriate device's deviceId.
+    Zoom is also a boolean constrainable property corresponding to the camera's ability to zoom (which results in a proper non-degenerate numeric setting range) and to the user agent's decision to expose that ability.
+    This boolean constraint is satisfied by a <a>settings dictionary</a> if the constraint value matches the existence of the numeric zoom setting in the settings dictionary.
```
Comment (jan-ivar) [inline index.bs:815]:
```suggestion
    Constraints on zoom influence camera selection through {{fitness distance}} toward cameras with the ability to zoom. To exert this influence without overwriting the current zoom setting, zoom may be constrained to true. Conversely, constraining it to false disfavors cameras with the ability to zoom.
```

File: index.bs [hunk 4]
```diff
@@ -795,20 +789,19 @@ When the {{getSettings()}} method is invoked on a video stream track, the user a
...
+    Pan is also a boolean constrainable property corresponding to the camera's ability to pan (which results in a proper non-degenerate numeric setting range) and to the user agent's decision to expose that ability.
+    This boolean constraint is satisfied by a <a>settings dictionary</a> if the constraint value matches the existence of the numeric pan setting in the settings dictionary.
 
-    If the {{visibilityState}} of the <a>top-level browsing context</a> value is "hidden", the {{applyConstraints()}} algorithm MUST throw a {{SecurityError}} if {{MediaTrackConstraintSet/pan}} dictionary member exists after a possible normalization.
+    Any algorithm which uses a {{MediaTrackConstraintSet}} object whose {{MediaTrackConstraintSet/pan}} dictionary member exists MUST either <a>request permission to use</a> (as defined in [[!permissions]]) a PermissionDescriptor with its name member set to {{PermissionName/camera}} and its {{CameraDevicePermissionDescriptor/panTiltZoom}} member set to true, and, optionally, consider its {{DevicePermissionDescriptor/deviceId}} member set to any appropriate device's deviceId or decide not to expose the pan setting.
```
Comment (jan-ivar) [inline index.bs:795]:
```suggestion
    Any algorithm which uses a {{MediaTrackConstraintSet}} object whose {{MediaTrackConstraintSet/pan}} dictionary member exists with a value other than false MUST either <a>request permission to use</a> (as defined in [[!permissions]]) a PermissionDescriptor with its name member set to {{PermissionName/camera}} and its {{CameraDevicePermissionDescriptor/panTiltZoom}} member set to true, and, optionally, consider its {{DevicePermissionDescriptor/deviceId}} member set to any appropriate device's deviceId, or decide not to expose the pan setting.
```

File: index.bs [hunk 5]
```diff
@@ -795,20 +789,19 @@ When the {{getSettings()}} method is invoked on a video stream track, the user a
...
+    Tilt is also a boolean constrainable property corresponding to the camera's ability to tilt (which results in a proper non-degenerate numeric setting range) and to the user agent's decision to expose that ability.
+    This boolean constraint is satisfied by a <a>settings dictionary</a> if the constraint value matches the existence of the numeric tilt setting in the settings dictionary.
+
+    Any algorithm which uses a {{MediaTrackConstraintSet}} object whose {{MediaTrackConstraintSet/tilt}} dictionary member exists MUST either <a>request permission to use</a> (as defined in [[!permissions]]) a PermissionDescriptor with its name member set to {{PermissionName/camera}} and its {{CameraDevicePermissionDescriptor/panTiltZoom}} member set to true, and, optionally, consider its {{DevicePermissionDescriptor/deviceId}} member set to any appropriate device's deviceId or decide not to expose the tilt setting.
```
Comment (jan-ivar) [inline index.bs:804]:
```suggestion
    Any algorithm which uses a {{MediaTrackConstraintSet}} object whose {{MediaTrackConstraintSet/tilt}} dictionary member exists with a value other than false MUST either <a>request permission to use</a> (as defined in [[!permissions]]) a PermissionDescriptor with its name member set to {{PermissionName/camera}} and its {{CameraDevicePermissionDescriptor/panTiltZoom}} member set to true, and, optionally, consider its {{DevicePermissionDescriptor/deviceId}} member set to any appropriate device's deviceId, or decide not to expose the tilt setting.
```

File: index.bs [hunk 6]
```diff
@@ -818,10 +811,12 @@ When the {{getSettings()}} method is invoked on a video stream track, the user a
...
+    Zoom is also a boolean constrainable property corresponding to the camera's ability to zoom (which results in a proper non-degenerate numeric setting range) and to the user agent's decision to expose that ability.
+    This boolean constraint is satisfied by a <a>settings dictionary</a> if the constraint value matches the existence of the numeric zoom setting in the settings dictionary.
 
-    If the {{visibilityState}} of the <a>top-level browsing context</a> value is "hidden", the {{applyConstraints()}} algorithm MUST throw a {{SecurityError}} if {{MediaTrackConstraintSet/zoom}} dictionary member exists after a possible normalization.
+    Any algorithm which uses a {{MediaTrackConstraintSet}} object whose {{MediaTrackConstraintSet/zoom}} dictionary member exists MUST either <a>request permission to use</a> (as defined in [[!permissions]]) a PermissionDescriptor with its name member set to {{PermissionName/camera}} and its {{CameraDevicePermissionDescriptor/panTiltZoom}} member set to true, and, optionally, consider its {{DevicePermissionDescriptor/deviceId}} member set to any appropriate device's deviceId or decide not to expose the zoom setting.
```
Comment (jan-ivar) [inline index.bs:817]:
```suggestion
    Any algorithm which uses a {{MediaTrackConstraintSet}} object whose {{MediaTrackConstraintSet/zoom}} dictionary member exists with a value other than false MUST either <a>request permission to use</a> (as defined in [[!permissions]]) a PermissionDescriptor with its name member set to {{PermissionName/camera}} and its {{CameraDevicePermissionDescriptor/panTiltZoom}} member set to true, and, optionally, consider its {{DevicePermissionDescriptor/deviceId}} member set to any appropriate device's deviceId, or decide not to expose the zoom setting.
```

**Review by jan-ivar — CHANGES_REQUESTED** [review]
Logic seems sound, but once w3c/mediacapture-main#766 is merged, I'm hoping we can reduce the prose even more. PTAL

**Comment by jan-ivar** [comment]
We should get this PR in before wide review of image capture starts, since it takes advantage of stuff we merged in https://github.com/w3c/mediacapture-main/pull/766.

To that end, I'll commit my suggestions above to this PR and solicit reviews.

File: index.bs [hunk 7]
```diff
@@ -795,20 +789,17 @@ When the {{getSettings()}} method is invoked on a video stream track, the user a
...
 
-    <div class="note">
-      The {{MediaTrackConstraintSet/pan}}, {{MediaTrackConstraintSet/tilt}} and {{MediaTrackConstraintSet/zoom}} dictionary members exists after a possible normalization if the normalized value is a double value or a {{ConstrainDoubleRange}} value (whether empty or not).
-      A boolean value of true is normalized to an empty {{ConstrainDoubleRange}} value thus the dictionary member exists after a normalization.
-      A boolean value of false is normalized to no value thus the dictionary member does not exist after a normalization.
-      An empty {{ConstrainDoubleRange}} value implies no constraints but only a permission and capability request.
-    </div></li>
+    If the {{visibilityState}} of the <a>top-level browsing context</a> value is "hidden", the {{applyConstraints()}} algorithm MUST throw a {{SecurityError}} if {{MediaTrackConstraintSet/pan}} dictionary member exists after a possible normalization.</li>
```
Comment (jan-ivar) [inline index.bs:796]:
```suggestion
    If the {{visibilityState}} of the <a>top-level browsing context</a> value is "hidden", the {{applyConstraints()}} algorithm MUST throw a {{SecurityError}} if {{MediaTrackConstraintSet/pan}} dictionary member exists with a value other than false.</li>
```

File: index.bs [hunk 8]
```diff
@@ -795,20 +789,17 @@ When the {{getSettings()}} method is invoked on a video stream track, the user a
...
+    Constraints on tilt influence camera selection through {{fitness distance}} toward cameras with the ability to tilt. To exert this influence without overwriting the current tilt setting, tilt may be constrained to true. Conversely, constraining it to false disfavors cameras with the ability to tilt.
+
+    Any algorithm which uses a {{MediaTrackConstraintSet}} object whose {{MediaTrackConstraintSet/tilt}} dictionary member exists with a value other than false MUST either <a>request permission to use</a> (as defined in [[!permissions]]) a PermissionDescriptor with its name member set to {{PermissionName/camera}} and its {{CameraDevicePermissionDescriptor/panTiltZoom}} member set to true, and, optionally, consider its {{DevicePermissionDescriptor/deviceId}} member set to any appropriate device's deviceId, or decide not to expose the tilt setting.
 
     If the {{visibilityState}} of the <a>top-level browsing context</a> value is "hidden", the {{applyConstraints()}} algorithm MUST throw a {{SecurityError}} if {{MediaTrackConstraintSet/tilt}} dictionary member exists after a possible normalization.
```
Comment (jan-ivar) [inline index.bs:804]:
```suggestion
    If the {{visibilityState}} of the <a>top-level browsing context</a> value is "hidden", the {{applyConstraints()}} algorithm MUST throw a {{SecurityError}} if {{MediaTrackConstraintSet/tilt}} dictionary member exists with a value other than false.
```

File: index.bs [hunk 9]
```diff
@@ -818,10 +809,11 @@ When the {{getSettings()}} method is invoked on a video stream track, the user a
...
 
-    If the {{visibilityState}} of the <a>top-level browsing context</a> value is "hidden", the {{applyConstraints()}} algorithm MUST throw a {{SecurityError}} if {{MediaTrackConstraintSet/zoom}} dictionary member exists after a possible normalization.
+    Any algorithm which uses a {{MediaTrackConstraintSet}} object whose {{MediaTrackConstraintSet/zoom}} dictionary member exists with a value other than false MUST either <a>request permission to use</a> (as defined in [[!permissions]]) a PermissionDescriptor with its name member set to {{PermissionName/camera}} and its {{CameraDevicePermissionDescriptor/panTiltZoom}} member set to true, and, optionally, consider its {{DevicePermissionDescriptor/deviceId}} member set to any appropriate device's deviceId, or decide not to expose the zoom setting.
 
+    If the {{visibilityState}} of the <a>top-level browsing context</a> value is "hidden", the {{applyConstraints()}} algorithm MUST throw a {{SecurityError}} if {{MediaTrackConstraintSet/zoom}} dictionary member exists after a possible normalization.</li>
```
Comment (jan-ivar) [inline index.bs:816]:
```suggestion
    If the {{visibilityState}} of the <a>top-level browsing context</a> value is "hidden", the {{applyConstraints()}} algorithm MUST throw a {{SecurityError}} if {{MediaTrackConstraintSet/zoom}} dictionary member exists with a value other than false.</li>
```

**Review by jan-ivar — COMMENTED** [review]
s/after a possible normalization/with a value other than false/

**Comment by jan-ivar** [comment]
@riju @youennf PTAL.

TL;DR: We no longer need to normalize away boolean inputs to non-boolean constraints as they're handled in [fitness distance](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-fitness-distance): _"4. If constraintValue is a boolean, but the constrainable property is not, then the fitness distance is based on whether the settings dictionary's constraintName member exists or not, from the formula `(constraintValue == exists) ? 0 : 1`"_

**Review by riju — APPROVED** [review]
Thanks @eehakkin, I think all of @jan-ivar 's requested changed are now handled.

**Comment by eehakkin** [comment]
Thanks @jan-ivar. As a final touch, I replaced occurrences of `{{fitness distance}}` with `<a>fitness distance</a>` so that links work properly.
